src/artifact_providers/github.ts

```diff
@@ -216,3 +235,17 @@ export class GitHubArtifactProvider extends BaseArtifactProvider {
     return await this.downloadAndUnpackArtifacts(archiveUrl);
   }
 }
+
+export function lazyRequest<T>(cb: lazyRequestCallback<T>): lazyRequestCallback<T> {
+  let data: T;
+  return async () => {
+    if (!data) {
+      data = await cb();
+    }
+    return data;
+  }
+}
```

Comment:
This looks a lot like Promise caching?

Reply:
Yeah that's what it's doing, let me know if there is a cleaner approach and / or naming suggestions.

Reply:
You can `await` a promise multiple times so I don't really understand the use case here.

You can do something like:
```js
const commitPromise = 
const getRevisionDate = async () => {
  return (await commitPromise).data.committer.date;
}
```

I also don't see the value of delaying the request until we are in the for loop as the for loop _will_ happen at least once, so why not just fetch the revision date once, before the loop and use that after on, eliminating another async step in the for loop (which should be _slightly_ more efficient):

```js
const revisionDate = await this.github.git.getCommit({
  owner,
  repo,
  commit_sha: revision,
}).data.committer.date;
```

Reply:
the request for the commit date is delayed because most of the time it isn't needed (the first page will _always_ be requested)

Reply:
Correct. That said even in that case I'm not sure if this generalized `lazyRequest` interface that is only being used in a single place is still needed.

Since `searchForRevisionArtifact` is already passed `revision` and accesses the API for other things, it can encapsulate the delayed request and then caching its response in itself, making both testing and following code easier.

Reply:
sure sure -- was mostly addressing this bit of your comment:

> why not just fetch the revision date once, before the loop and use that after on

Reply:
`searchForRevisionArtifact` may be called multiple times `getRevisionArtifact`.

The options for managing getting the commit is:

1. Get the commit in each `searchForRevision` call (can result in three calls vs one) and is what I had in the first commit for this PR.
2. Get the commit once in `getRevisionArtifact` and pass it to `searchForRevision` which may result in an unnecessary API call where the artifact is on the first page and the revision commit is never queried. (This is a small issue but was the original behavior)
3. The lazy request approach where the API call is made when needed and data is cached.

With respect to this vs the promise caching, I think that'll run into option 2 of making a request when we may no need it.

Reply:
Yeah, I see the reasoning. Was just arguing that the lazy fetch approach is adding some complexity into the code that is also not used anywhere to just save a single API request, that's all.

Reply:
that complexity was already existing -- it's now just preserved over retries